src/native/libs/System.Net.Security.Native/pal_gssapi.c

```diff
@@ -706,6 +717,52 @@ uint32_t NetSecurityNative_IsNtlmInstalled(void)
     return foundNtlm;
 }
 
+uint32_t NetSecurityNative_InquireSecContextSessionKey(uint32_t* minorStatus,
+                                                       GssCtxId* contextHandle,
+                                                       PAL_GssBuffer* outBuffer)
+{
+    assert(minorStatus != NULL);
+    assert(contextHandle != NULL);
+    assert(outBuffer != NULL);
+
+    uint32_t minorStatusFree;
+    size_t i;
+#if HAVE_GSS_C_INQ_SSPI_SESSION_KEY
+    gss_OID desiredObject = GSS_C_INQ_SSPI_SESSION_KEY;
+#else
+    gss_OID desiredObject = &gss_c_inq_sspi_session_key_OID_desc;
+#endif
+    gss_buffer_set_t sessionKey = GSS_C_NO_BUFFER_SET;
+
+    uint32_t majorStatus =
+        gss_inquire_sec_context_by_oid(minorStatus, contextHandle, desiredObject, &sessionKey);
+
+    if (majorStatus != GSS_S_COMPLETE)
+    {
+        return majorStatus;
+    }
+
+    // We expect at least 1 buffer in the set with the first being the session key.
+    if (sessionKey->count < 1)
+    {
+        gss_release_buffer_set(&minorStatusFree, &sessionKey);
+        return GSS_S_FAILURE;
+    }
+
+    // It is difficult to map the buffer set to the .NET shim so we copy the session key into a single buffer that can
+    // be freed by our shim. We need to make sure we clear out the remaining data set values and the set itself.
+    NetSecurityNative_MoveBuffer(&sessionKey->elements[0], outBuffer);
+
+    for (i = 1; i < sessionKey->count; i++)
+    {
+        gss_release_buffer(&minorStatusFree, &sessionKey->elements[i]);
+    }
+    free(sessionKey->elements);
+    free(sessionKey);
+
+    return majorStatus;
```

Comment:
Ideally it would be nice to return the buffer set and have the caller extra the element needed but the use of the shim makes it quite hard to store both the set structure and the elements in there. In the end I found it easier to just set the output buffer to the one that's needed and manually free the remaining buffer set values that `gss_release_buffer_set` would have done for us anyway.